### ISSUE TITLE
Implement CRF and Viterbi decoding  by some elementary operators

### ISSUE THREAD

**Issue by lcy-seso** [PaddlePaddle/Paddle#4535]
The computation of CRF, in essence, can be regarded as "softmax with cross entropy", but the normalization factor is more complicated.

Maybe, in the near feature, with a better support of scatter/gather/ conditional operators, we can also try to implement CRF and the Viterbi decoding process by using elementary operators (**with no sequence padding**) to test our framework.

Just my personal consideration, maybe it is not very necessary.

**Comment by lcy-seso** [comment]
And another amazing thing is CRF can be regarded as an RNN. Potentially, it may be implemented in a similar way we treat RNN. 

I am not very sure yet but found CRF can be implemented in a much better design than I have ever thought.
